File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.cpp
class Solution {
public:
    int firstStableIndex(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> suffix(n,0);

        suffix[n-1] = nums[n-1];

        for(int i = n-2;i>=0;i--){
            suffix[i] = min(nums[i],suffix[i+1]);
        }

        int maxi = INT_MIN;
        for(int i = 0 ;i < n;i++){
            maxi = max(nums[i],maxi);
            if((maxi - suffix[i]) <= k)return i;
        }

        return -1;
    }
};

File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        n = len(nums)

        suff = [1]*n
        suff[n-1] = nums[n-1]

        for i in range(n-2,-1,-1):
            suff[i] = min(nums[i],suff[i+1])

        maxi = nums[0]
        for i in range(len(nums)):
            maxi = max(nums[i],maxi)
            if maxi - suff[i] <= k: return i
        
        return -1